cpmpy/transformations/linearize.py

```diff
@@ -445,6 +435,47 @@ def only_positive_bv(lst_of_expr):
 
     return newlist
 
+def only_positive_bv_sub(expr):
+    """
+        Replaces a linear expression containing NegBoolView with an equivalent expression using only BoolVar.
+        expr is expected to be a var, sum or wsum.
+        
+        If the expression has a NegBoolView, the function returns a tuple of the new expression as a wsum and a constant, 
+        which can be added to the wsum or subtracted from the rhs (depending on the usecase).
+        Otherwise the original expression is returned with a constant of 0.
+    """
+    if isinstance(expr, _NumVarImpl):
+        if isinstance(expr,NegBoolView):
+            expr, const = Operator("wsum",[[-1], [expr._bv]]), 1
+        else:
+            expr, const = expr, 0
+
+    elif expr.name == "sum":
+        if any(isinstance(a, NegBoolView) for a in expr.args):
+        # count number of negboolviews
+            const = sum([1 for a in expr.args if isinstance(a, NegBoolView)])
+            expr = Operator("wsum",[[-1 if isinstance(arg, NegBoolView) else 1 for arg in expr.args], [arg._bv if isinstance(arg, NegBoolView) else arg for arg in expr.args]])
+        else:
+            expr, const = expr, 0
+
+    elif expr.name == "wsum":
+        weights, args = expr.args
+        if any(isinstance(a, NegBoolView) for a in args):
+            idxes = {i for i, a in enumerate(args) if isinstance(a, NegBoolView)}
```

Comment:
this is more in the direction of what I suggested, but you are doing it after the 'any' that already loops over everything, and you are using a dict which seems wastefull here... you can just have a boolean vector that you check on in your loops...